default/python/universe_generation/empires.py

```diff
@@ -95,26 +85,101 @@ def min_planets_in_vicinity_limit(num_systems):
+
+        # In the list of systems reverse sorted by merit, if the first
+        # num_home_systems are all min_jumps apart, then the merit of the
+        # num_home_systems nth system is the best_case_merit_lower_bound that
+        # can be achieved.
+        all_merit_system = sorted([(self.system_merit[s], s)
+                                   for s in systems_pool], reverse=True)
+        current_merit_lower_bound = 0
+
+        best_candidate = []
+
+        # cap the number of attempts to the smaller of the number of systems in the pool, or 100
+        attempts = min(100, len(systems_pool))
+
+        # Cap the number of attempts when the found number of systems is much less than the target
+        # num_home_systems because this indicates that the min_jumps is too large and/or the
+        # systems_pool is too small to ever succeed.
+        num_complete_misses_remaining = 4
+
+        while attempts and num_complete_misses_remaining:
+            # use a local pool of all candidate systems better than the worst threshold merit
+            all_merit_system = [(m, s) for (m, s) in all_merit_system if m > current_merit_lower_bound]
+            local_pool = {s for (m, s) in all_merit_system}
+
+            if len(local_pool) < self.num_home_systems:
+                if len(best_candidate) < self.num_home_systems:
+                    print ("Failing in find_home_systems_for_min_jump_distance because "
+                           "current_merit_lower_bound = {} trims local pool to {} systems "
+                           "which is less than num_home_systems {}.".format(
+                               current_merit_lower_bound, len(local_pool), self.num_home_systems))
+                break
+
+            attempts = min(attempts - 1, len(local_pool))
+
+            candidate = []
+            while local_pool:
+                member = random.choice(list(local_pool))
+                candidate.append(member)
+
+                # remove all neighbors from the local pool
+                local_pool -= set(get_systems_within_jumps(member, min_jumps))
+
+            # Count complete misses when number of candidates is not close to the target.
+            if len(candidate) < (self.num_home_systems - 4):
+                num_complete_misses_remaining -= 1
+
+            if len(candidate) < self.num_home_systems:
+                continue
+
+            # Calculate the merit of the current attempt.  If it is the best so far
+            # keep it and update the merit_threshold
+            merit_system = sorted([(self.system_merit[s], s)
+                                   for s in candidate], reverse=True)[:self.num_home_systems]
+
+            (merit, system) = merit_system[-1]
+
+            # Quit if the lowest merit planet meets the minimum threshold
+            if merit >= min_planets_in_vicinity_limit(get_systems_within_jumps(system, HS_VICINITY_RANGE)):
+                break
```

Comment:
Shouldn't that happen *after* checking if `merit` is better than `current_merit_lower_bound`, and `best_candidate` has been updated? Because doing that *here* will result in the candidate under consideration in the current iteration being discarded, regardless if it's better than the current `best_candidate` or not.

Which can, worst case, lead to quitting before *any* best candidate has been determined, when the candidate of the first iteration already meets this exit criteria. Resulting in returning an empty `best_candidate`, definitely not what we want. Unless I missed something obvious of course...

That check only makes sense if a new best candidate has been found anyway, doesn't it? Otherwise the loop would have already been quit when the last best candidate had been found.